### PR TITLE
Add rpi-gpio types

### PR DESCRIPTION
**Pull request by Roaders**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [x] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/rpi-gpio/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+/// <reference types="node" />
+import { EventEmitter } from 'events';
+
+type MODE = 'mode_rpi' | 'mode_bcm';
+type DIR = 'in' | 'out' | 'low' | 'high';
+type EDGE = 'none' | 'rising' | 'falling' | 'both';
+type ValueCallback<T> = (err?: Error | null, value?: T) => void;
+type ErrorCallback = (err?: Error | null) => void;
+
+declare class Gpio extends EventEmitter {
```
Comment (peterblazejewicz) [inline types/rpi-gpio/index.d.ts:16]: for your consideration, should it be exposed as the class that someone can create? The API seems to be always an already created instance, so here one could expect abstract or interface to describe the contract:
https://github.com/JamesBarwell/rpi-gpio.js/blob/6dd19bf1adf8ba3ac48f13bdc1aa60f4d05341ee/rpi-gpio.js#L507

File: types/rpi-gpio/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+     * Set pin reference mode. Defaults to 'mode_rpi'.
+     *
+     * @param mode Pin reference mode, 'mode_rpi' or 'mode_bcm'
+     */
+    setMode(mode: MODE): void;
+    /**
+     * Setup a channel for use as an input or output
+     *
+     * @param  channel   Reference to the pin in the current mode's schema
+     * @param _direction The pin direction, either 'in' or 'out'
```
Comment (peterblazejewicz) [inline types/rpi-gpio/index.d.ts:39]: out of the curiosity, why this convention with lodash? it's very uncommon

File: types/rpi-gpio/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+// Project: https://github.com/JamesBarwell/rpi-gpio.js#readme
+// Definitions by: Giles Roadnight <https://github.com/Roaders>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// Minimum TypeScript Version: 3.0
+
+/// <reference types="node" />
+import { EventEmitter } from 'events';
+
+type MODE = 'mode_rpi' | 'mode_bcm';
+type DIR = 'in' | 'out' | 'low' | 'high';
```
Comment (peterblazejewicz) [inline types/rpi-gpio/index.d.ts:11]: there is really no reason why this cannot be a full word with explicit meaning
```suggestion
type PinDirection = 'in' | 'out' | 'low' | 'high';
```
https://github.com/JamesBarwell/rpi-gpio.js#setupchannel--direction-edge-callback
note that author of package seems to use explicit terms:
https://github.com/JamesBarwell/rpi-gpio.js/blob/6dd19bf1adf8ba3ac48f13bdc1aa60f4d05341ee/rpi-gpio.js#L90-L94

File: types/rpi-gpio/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+     */
+    setup(channel: number, onSetup?: ValueCallback<boolean>): void;
+    setup(channel: number, direction?: DIR, onSetup?: ValueCallback<boolean>): void;
+    setup(channel: number, direction?: DIR, edge?: EDGE, onSetup?: ValueCallback<boolean>): void;
+    /**
+     * Write a value to a channel
+     *
+     * @param channel The channel to write to
+     * @param value   If true, turns the channel on, else turns off
+     * @param cb      Optional callback
```
Comment (peterblazejewicz) [inline types/rpi-gpio/index.d.ts:51]:
```suggestion
     * @param [cb]      callback
```
JSDoc convention honored by TS

File: types/rpi-gpio/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+     */
+    write(channel: number, value: boolean, cb?: ErrorCallback): void;
+    /**
+     * Write a value to a channel
+     *
+     * @param channel The channel to write to
+     * @param value   If true, turns the channel on, else turns off
+     * @param cb      Optional callback
```
Comment (peterblazejewicz) [inline types/rpi-gpio/index.d.ts:59]: ditto:
```suggestion
     * @param [cb]      callback
```

File: types/rpi-gpio/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+     * @param channel The channel to read from
+     * @param cb      Callback which receives the channel's boolean value
+     */
+    input(channel: number, cb: ValueCallback<boolean>): void;
+    /**
+     * Unexport any pins setup by this module
+     *
+     * @param cb Optional callback
+     */
+    destroy(cb: ErrorCallback): void;
```
Comment (peterblazejewicz) [inline types/rpi-gpio/index.d.ts:81]: the api describes this as no args one:
https://github.com/JamesBarwell/rpi-gpio.js#destroy
but the source as optional:
https://github.com/JamesBarwell/rpi-gpio.js/blob/master/rpi-gpio.js#L310
```suggestion
    destroy(cb?: ErrorCallback): void;
```

File: types/rpi-gpio/index.d.ts [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+     * Read a value from a channel
+     *
+     * @param channel The channel to read from
+     * @param cb      Callback which receives the channel's boolean value
+     */
+    input(channel: number, cb: ValueCallback<boolean>): void;
+    /**
+     * Unexport any pins setup by this module
+     *
+     * @param cb Optional callback
```
Comment (peterblazejewicz) [inline types/rpi-gpio/index.d.ts:79]: ditto:
```suggestion
     * @param [cb] callback
```

File: types/rpi-gpio/index.d.ts [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+    private setRaspberryVersion;
+    private getPinRpi;
+    private getPinBcm;
+    /**
+     * Listen for interrupts on a channel
+     *
+     * @param channel The channel to watch
+     * @param cb Callback which receives the channel's err
+     */
+    private listen;
```
Comment (peterblazejewicz) [inline types/rpi-gpio/index.d.ts:95]: are those required? should be here?
```suggestion

```

**Review by peterblazejewicz — CHANGES_REQUESTED** [review]
@Roaders looks OK, minor topics to review, otherwise LGTM!

File: types/rpi-gpio/index.d.ts (hunk 1 above)
Comment (Roaders) [reply types/rpi-gpio/index.d.ts:16]: I'm just following what the javascript does. The class is not exported by the javascript so the class won't be available to instantiate.

File: types/rpi-gpio/index.d.ts (hunk 2 above)
Comment (Roaders) [reply types/rpi-gpio/index.d.ts:39]: this was internal implementation when I was working on a TS version of this library (the basis of these types). I'll remove

File: types/rpi-gpio/index.d.ts (hunk 6 above)
Comment (Roaders) [reply types/rpi-gpio/index.d.ts:81]: In this case the source is incorrect. If you do pass a callback you get errors. I will correct the docs comment

**Review by peterblazejewicz — APPROVED** [review]
LGTM!
@Roaders thx for the contribution!
I'm on the Pi side too, thanks!

**Comment by Roaders** [comment]
Is this PR good to go then? Do I need to do anything else to get it merged?

**Comment by peterblazejewicz** [comment]
Yes, ofc, let me bring the cavalry!

File: types/rpi-gpio/rpi-gpio-tests.ts [hunk 9]
```diff
@@ -0,0 +1,36 @@
+import gpio from "rpi-gpio";
```
Comment (sandersn) [inline types/rpi-gpio/rpi-gpio-tests.ts:1]: Please use the commonjs syntax `import gpio = require('rpi-gpio')`, and remove "esModuleInterop": true from tsconfig.

Consumers of the package who want to use `import gpio from 'rpi-gpio'` will need to add `"esModuleInterop": true`, but Definitely Typed tests need to be backward compatible to show that the types will work with old code bases.

**Review by sandersn — CHANGES_REQUESTED** [review]

**Review by sandersn — APPROVED** [review]

**Pull request merged**